File: xtracfg/cmd/store/check.go
package store

import (
	"errors"
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/interactive-instruments/xtraplatform-cli/libxtracfg/go/xtracfg"
	"github.com/interactive-instruments/xtraplatform-cli/xtracfg/util"
	"github.com/spf13/cobra"
)

var ignoreRedundant *bool

// CheckCmd represents the check command
func CheckCmd(store xtracfg.Store, name string, verbose *bool, debug *bool) *cobra.Command {
	check := &cobra.Command{
		Use:   "check",
		Short: "Check the store source",
		Long: `Check the store source
Executes all subcommands in order, see the subcommand help for details.`,
		Args: cobra.NoArgs,
		Run: func(cmd *cobra.Command, args []string) {
			if *debug {
				fmt.Fprint(os.Stdout, "Checking store source: ", store.Label(), "\n")
			}
			results, err := store.Handle(map[string]interface{}{"ignoreRedundant": strconv.FormatBool(*ignoreRedundant)}, "check")

			util.PrintResults(results, err)

			printFix(results, err, name)
		},
	}

	ignoreRedundant = check.PersistentFlags().BoolP("ignore-redundant", "r", false, "ignore redundant settings")

	check.PersistentFlags().SortFlags = false
	check.Flags().SortFlags = false

	checkCfg := &cobra.Command{
		Use:   "cfg",
		Short: "Check cfg.yml in the store source",
		Long:  `Checks cfg.yml for deprecated settings.`,
		Args:  cobra.NoArgs,
		Run: func(cmd *cobra.Command, args []string) {
			if *debug {
				fmt.Fprint(os.Stdout, "Checking cfg.yml in the store source: ", store.Label(), "\n")
			}

			results, err := store.Handle(map[string]interface{}{"ignoreRedundant": strconv.FormatBool(*ignoreRedundant)}, "check", "cfg")

			util.PrintResults(results, err)

			printFix(results, err, name)
		},
	}

	checkEntities := &cobra.Command{
		Use:   "entities [path]",
		Short: "Check entities in the store source",
		Long: `Checks entity configurations for deprecated, unknown and redundant settings.
To check only a single entity, pass the path to the file relative to the source as argument.`,
		Example: name + " check entities -v -r \n" + name + " check entities -v -r store/entities/services/api.yml",
		Args: func(cmd *cobra.Command, args []string) error {
			if len(args) > 1 {
				return errors.New("only one argument expected")
			}
			//TODO
			isValidPath := true
			if isValidPath {
				return nil
			}
			return fmt.Errorf("invalid entity path specified: %s", args[0])
		},
		Run: func(cmd *cobra.Command, args []string) {
			if *debug {
				fmt.Fprint(os.Stdout, "Checking store source: ", store.Label(), "\n")
			}
			path := ""
			if len(args) > 0 {
				path = args[0]
			}

			results, err := store.Handle(map[string]interface{}{"ignoreRedundant": strconv.FormatBool(*ignoreRedundant), "path": path}, "check", "entities")

			util.PrintResults(results, err)

			printFix(results, err, name)
		},
	}

	checkLayout := &cobra.Command{
		Use:   "layout",
		Short: "Check layout of the store source",
		Long:  `Checks for a deprecated directory layout.`,
		Args:  cobra.NoArgs,
		Run: func(cmd *cobra.Command, args []string) {
			if *debug {
				fmt.Fprint(os.Stdout, "Checking layout of store source: ", store.Label(), "\n")
			}

			results, err := store.Handle(map[string]interface{}{"ignoreRedundant": strconv.FormatBool(*ignoreRedundant)}, "check", "layout")

			util.PrintResults(results, err)

			printFix(results, err, name)
		},
	}

	check.AddCommand(checkCfg)
	check.AddCommand(checkEntities)
	check.AddCommand(checkLayout)

	return check
}

func printFix(results []xtracfg.Result, err error, name string) {
	if err == nil && xtracfg.HasStatus(results, xtracfg.Warning) {
		fmt.Fprint(os.Stdout, "\n", "Run '", name, " ", strings.Replace(strings.Join(os.Args[1:], " "), "check", "upgrade", 1), "' to fix all detected issues.", "\n")
	} else {
		fmt.Fprint(os.Stdout, "\n")
	}
}

File: xtracfg/src/main/java/de/ii/xtraplatform/cli/cmd/Check.java
package de.ii.xtraplatform.cli.cmd;

import de.ii.ldproxy.cfg.Layout;
import de.ii.ldproxy.cfg.LdproxyCfg;
import de.ii.xtraplatform.cli.*;
import java.util.Map;
import java.util.Optional;

public class Check extends Store {

  public Check(Optional<String> subcommand, Map<String, Object> parameters) {
    super(subcommand, parameters);
  }

  public Result run(Subcommand cmd, LdproxyCfg ldproxyCfg, Layout layout) {
    switch (cmd) {
      case cfg:
        return CfgHandler.check(ldproxyCfg, ignoreRedundant, verbose, debug);
      case defaults:
        return EntitiesHandler.check(
            ldproxyCfg, EntitiesHandler.Type.Defaults, path, ignoreRedundant, verbose, debug);
      case entities:
        return EntitiesHandler.check(
            ldproxyCfg, EntitiesHandler.Type.All, path, ignoreRedundant, verbose, debug);
      case overrides:
        return Result.empty();
      case layout:
        return LayoutHandler.check(layout, verbose);
      default:
        throw new IllegalStateException("Unexpected subcommand: " + subcommand);
    }
  }

  @Override
  public Result run(Context context) {
    Result result = Result.empty();

    for (Subcommand cmd : Subcommand.values()) {
      if (shouldRun(cmd)) {
        result = result.merge(run(cmd, context.ldproxyCfg, context.layout));
      }
    }

    if (result.isEmpty()) {
      result.success("Everything is fine");
    }

    return result;
  }
}
